### ISSUE TITLE
Images/Screenshots

### ISSUE THREAD

**Issue by Jargonius** [heyman/heynote#157]
It would be nice if you could paste basic images and screenshots into blocks.  Sometimes it is quicker to take a screenshot than it is to copy something.

**Comment by heyman** [comment]
I agree that this would be useful. However, it would be quite a significant conceptual change. 

If it were to be implemented, these are my inital thoughts:

* It should only work in Text blocks (and potentially Math blocks), because it would be weird to paste images into programming code.
* Images could be stored in a sub-directory within the Heynote buffer directory, and then referenced with a special syntax (e.g. `<<image:attachments/2024-01-12T12:54:07+01:00_abcd123.png>>`)

### Questions

* When should images be deleted from the file system? When references are removed from the text? If so, what about cutting a piece of text (including an image) and pasting it somewhere else?
* How to handle copy and pasting of mixed text / image contents? It should be possible to copy and paste sections of both text and images within Heynote, but what about to/from external programs?
* What about Markdown blocks? There's a syntax for images in Markdown, but the referenced images would be broken when copying the content from Heynote out to an external program (e.g. Github comments). I'm thinking it might be better to *not* support image widgets in Markdown blocks.

**Comment by gerroon** [comment]
It would be even better HeyNote can do the screenshot operation. Does not Electron apps support screen shot?

**Comment by heyman** [comment]
> It would be even better HeyNote can do the screenshot operation.

Sorry, I don't understand what you mean by that. If you mean building a feature for taking screenshots/screen captures into Heynote, that's not going to happen.

**Comment by gerroon** [comment]
I did not mean implementing a new feature. I thought that Electron already has faculty to do screenshots.

https://www.codeease.net/programming/javascript/how-to-take-screenshot-of-desktop-using-electron-js

**Comment by peterjaric** [comment]
>It should only work in Text blocks (and potentially Math blocks), because it would be weird to paste images into programming code.

That begets a new question: what happens if you have pasted an image into a Text block and then change the language to Python, for example?

**Comment by Jargonius** [comment]
All of these are good points that I hadn't considered.  I'll have to think about this.  But anyone else who has ideas please chime in.  I think we've got a good dialogue going so far.  One way to simplify it would be to make a special image block that can't accept text or be changed and only deleted.  But that would reduce the versatility drastically.

**Comment by Jargonius** [comment]
What about base64 encoding the images?

**Comment by Jargonius** [comment]
> All of these are good points that I hadn't considered.  I'll have to think about this.  But anyone else who has ideas please chime in.  I think we've got a good dialogue going so far.
> 
> One way to simplify it would be to make a special image block that can't accept text or be changed and only deleted.  But that would reduce the versatility drastically.

You could also disable changing the language if the block contains an image.

**Comment by Jargonius** [comment]
> >It should only work in Text blocks (and potentially Math blocks), because it would be weird to paste images into programming code.
> 
> That begets a new question: what happens if you have pasted an image into a Text block and then change the language to Python, for example? 

You could also disable changing the language if the block contains an image.

**Comment by porridgewithraisins** [comment]
> * When should images be deleted from the file system? When references are removed from the text? If so, what about cutting a piece of text (including an image) and pasting it somewhere else?

I think you should make delete orphan images a checkbox in the settings page, should be sufficient. Can probably think about running the deletion at program start as well (but this would break people cmd+Xing something, accidentally cmd+Qing heynote, restarting it and finding their image gone)

> * What about Markdown blocks? There's a syntax for images in Markdown, but the referenced images would be broken when copying the content from Heynote out to an external program (e.g. Github comments). I'm thinking it might be better to _not_ support image widgets in Markdown blocks.

Please support it in markdown. It is reasonable to expect markdown copied from the local filesystem to not have the same image references in a github comment.

**Comment by antonio-veezoo** [comment]
Any updates on this? Not having images make the app too basic for me to use

**Comment by SlyFoxHayes** [comment]
What about a dedicated image block? With that thought, there could be dedicated audio block, dedicated video block ...

**Comment by heyman** [comment]
> What about a dedicated image block?

Yes, that's probably a good solution.

**Comment by Tukks** [comment]
I think the image block will be a great idea.
If you ctrl+V an image it will automatically create an image block after the current block.

**Comment by rublev** [comment]
+1 !!

**Comment by Kur0x** [comment]
add image block is very useful to me.(Now I have to use OneNote to store images.)

**Comment by Bit0r** [comment]
> > It should only work in Text blocks (and potentially Math blocks), because it would be weird to paste images into programming code.
> 
> That begets a new question: what happens if you have pasted an image into a Text block and then change the language to Python, for example?

This shouldn't be a problem. In VS Code, you can also change Markdown to Python, and then VS Code will mark it with a red wavy line to indicate that it's a bunch of erroneous Python code.

**Comment by heyman** [comment]
I've started working on this (#429). Still quite a bit to go, but it feels promising so far.



>> What about a dedicated image block?
>
> Yes, that's probably a good solution.

I've settled on supporting images in all kinds of blocks. If we were to only have dedicated image blocks, it wouldn't solve the use-case of drafting an email (or something else) with images in a Heynote block and then being able to press `C-a C-c` to copy the whole block (including both text and images).

**Comment by heyman** [comment]
I've now released a new beta ([2.8.0-beta.2](https://github.com/heyman/heynote/releases/tag/v2.8.0-beta.2)) with image support and could use your help testing it. I'd love feedback on both the overall UI/UX, as well as ironing out any potential bugs.

Easiest way to install it is to turn on "Use beta versions of Heynote" in settings, and then press the "Check for updates" button in the bottom right corner:

This settings:
<img width="441" height="251" alt="Image" src="https://github.com/user-attachments/assets/c1787083-672b-4aaa-ad8f-c0b38e5b0270" />

The arrows circle in the corner:
<img width="98" height="45" alt="Image" src="https://github.com/user-attachments/assets/ac98fa26-7cb7-4d88-b18f-88f204265d3b" />

**Comment by ToxMox** [comment]
Just tested the images feature in 2.8.0-beta.3.
I have my heynote configured to use a custom location for the buffer files on my unraid server. I'm testing on Windows. 
When I add an image it properly copies the file to the storage location but the image shows up as a dead link. 
I tested with both UNC path and mapped drive. I was able to workaround the issue by disabling the Buffer Files Path custom location and instead renaming the notes folder in AppData to something else then making  a symlink to the unc path. That worked for me.

**Comment by theowenyoung** [comment]
I tested the latest version 2.8.0-beta.4 on macOS and it's working well! I was able to drag images into the app. However, I found the interaction a bit unintuitive at first. I initially tried copy-pasting images, which didn't work no matter what I tried. So I went to check the source code and found that it seems only drag-and-drop is currently supported.

Would you consider adding copy-paste support for images in the future? It would make the workflow more intuitive for users who are accustomed to using Cmd+C/Cmd+V for image handling.

Thanks for the great work on this feature!

**Comment by heyman** [comment]
> Just tested the images feature in 2.8.0-beta.3. I have my heynote configured to use a custom location for the buffer files on my unraid server. I'm testing on Windows. When I add an image it properly copies the file to the storage location but the image shows up as a dead link. I tested with both UNC path and mapped drive. I was able to workaround the issue by disabling the Buffer Files Path custom location and instead renaming the notes folder in AppData to something else then making a symlink to the unc path. That worked for me.

Could you check if the following build fixes it (scroll down to "Artifacts"):
https://github.com/heyman/heynote/actions/runs/21478608266

**Comment by heyman** [comment]
> Would you consider adding copy-paste support for images in the future? It would make the workflow more intuitive for users who are accustomed to using Cmd+C/Cmd+V for image handling.

I agree. Unfortunately Electron doesn't provide a cross-platform API for reading copied files (or their paths) from the clipboard, so I think we would need to write three different implementations for mac/windows/linux. Ultimately I'd like to support this but at the moment I have other improvements on the immediate todo list that I prioritize higher.

Note that copying image *data* and pasting into Heynote works right now.

**Comment by ToxMox** [comment]
> > Just tested the images feature in 2.8.0-beta.3. I have my heynote configured to use a custom location for the buffer files on my unraid server. I'm testing on Windows. When I add an image it properly copies the file to the storage location but the image shows up as a dead link. I tested with both UNC path and mapped drive. I was able to workaround the issue by disabling the Buffer Files Path custom location and instead renaming the notes folder in AppData to something else then making a symlink to the unc path. That worked for me.
> 
> Could you check if the following build fixes it (scroll down to "Artifacts"): https://github.com/heyman/heynote/actions/runs/21478608266

Seems to work well now, thanks!
